hubblestack/extmods/fdg/cert_discovery.py

```diff
@@ -0,0 +1,168 @@
+# -*- encoding: utf-8 -*-
+"""
+Flexible Data Gathering: cert_discovery
+=============================
+Intention -
+This fdg module allows connecting to open ports on a system and retrieving
+certificate details that might be attached on those ports.
+
+Testing -
+    1. Configure this module through Hubble's schedule using the following configuration
+        fdg_cert_discovery:
+        function: fdg.fdg
+        seconds: __seconds__
+        splay: __splay__
+        returner: __returner__
+        run_on_start: __run_on_start__
+        args:
+          - <path to fdg profile>
+    Make sure that cert_discovery.fdg profile exists and contains Osquery as the first
+    module and this module as the second module.
+    2. Alternately, execute hubble fdg.fdg <path to fdg profile> to run this module via cmd.
+"""
+import OpenSSL
+import ssl
+from socket import setdefaulttimeout
+import logging
+from datetime import datetime
+log = logging.getLogger(__name__)
+
+setdefaulttimeout(3)
```

Comment:
Default timeout should be made configurable either via profile or config. Adding this option in profile would make it more flexible. If this flag is not set explicitly anywhere in profile or config and then we can set it to some default like `3`.

Reply:
Actually this is not a user supplied value. This is more a preventive action on our side to exit the code if it gets stuck. We don't want the user to either supply or edit the value.

Reply:
User would never specify timeout duration. It would be Admins who manage Hubble profiles who should have flexibility to override default timeout.

Reply:
Remember too that Hubble is single threaded, so if you're checking 10 hosts ... and for argument's sake, let's say they're all super busy; and your timeout is say 15s ...

Then during the schedule item (fdg) we'd check the 10 hosts at 15s/check giving 150s for just that one schedule item.

(For reference, we have this problem in the HEC too... There we decided several failures in a row should flag an outage condition and simply queue to disk.)

Reply:
Configured timeout via the corresponding fdg profile.